### PR TITLE
 Remove the C++ test dependency on dig and dnsutils

### PR DESCRIPTION
**Pull request by apolcyn**
based on initial clean up PR https://github.com/grpc/grpc/pull/14609 (only the last commit is new).

Fixes https://github.com/grpc/grpc/issues/13621

Right now,  https://github.com/grpc/grpc/issues/13621 is not working because the `dig` command is not available in the docker image supplied by foundry. This PR fixes that by replacing usage of the `dig` command in tests with python scripts that do the equivalent.  

Alternatives include:
* using a custom docker image with the `dnsutils` package installed
* creating a bazel rule that compiles the `dig` command from source (this is an autoconf-based project)

But I believe this is the least painful option, and it probably helps for general test portability anyways.

cc @adelez

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: templates/test/cpp/naming/resolver_component_tests_defs.include [hunk 1]
```diff
@@ -41,10 +43,10 @@ echo "Local DNS server started. PID: $DNS_SERVER_PID"
 # Health check local DNS server TCP and UDP ports
 for ((i=0;i<30;i++));
 do
-  echo "Retry health-check DNS query to local DNS server over tcp and udp"
+  echo "Retry health-check local DNS server by attempting a DNS query and TCP handshake"
   RETRY=0
-  dig A health-check-local-dns-server-is-alive.resolver-tests.grpctestingexp. @localhost -p "$FLAGS_test_dns_server_port" +tries=1 +timeout=1 | grep '123.123.123.123' || RETRY=1
-  dig A health-check-local-dns-server-is-alive.resolver-tests.grpctestingexp. @localhost -p "$FLAGS_test_dns_server_port" +tries=1 +timeout=1 +tcp | grep '123.123.123.123' || RETRY=1
+  $FLAGS_test_dns_resolver_bin_path -s 127.0.0.1 -p "$FLAGS_test_dns_server_port" -n health-check-local-dns-server-is-alive.resolver-tests.grpctestingexp. -t 1 | grep '123.123.123.123' || RETRY=1
```
Comment (dgquintas) [inline templates/test/cpp/naming/resolver_component_tests_defs.include]: any ipv6 concerns? 127.0.0.1 is ipv4-specific. @pmarks-net

File: test/cpp/naming/resolver_component_tests_runner.sh [hunk 2]
```diff
@@ -41,10 +43,10 @@ echo "Local DNS server started. PID: $DNS_SERVER_PID"
 # Health check local DNS server TCP and UDP ports
 for ((i=0;i<30;i++));
 do
-  echo "Retry health-check DNS query to local DNS server over tcp and udp"
+  echo "Retry health-check local DNS server by attempting a DNS query and TCP handshake"
   RETRY=0
-  dig A health-check-local-dns-server-is-alive.resolver-tests.grpctestingexp. @localhost -p "$FLAGS_test_dns_server_port" +tries=1 +timeout=1 | grep '123.123.123.123' || RETRY=1
-  dig A health-check-local-dns-server-is-alive.resolver-tests.grpctestingexp. @localhost -p "$FLAGS_test_dns_server_port" +tries=1 +timeout=1 +tcp | grep '123.123.123.123' || RETRY=1
+  $FLAGS_test_dns_resolver_bin_path -s 127.0.0.1 -p "$FLAGS_test_dns_server_port" -n health-check-local-dns-server-is-alive.resolver-tests.grpctestingexp. -t 1 | grep '123.123.123.123' || RETRY=1
```
Comment (dgquintas) [inline test/cpp/naming/resolver_component_tests_runner.sh]: ditto wrt ipv4 vs ipv6

File: test/cpp/naming/test_dns_resolver.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  argp.add_argument('-n', '--qname', default=None, type=str,
+                    help=('Name of the record to query for. '))
+  argp.add_argument('-t', '--timeout', default=1, type=int,
+                    help=('Force process exit after this number of seconds.'))
+  args = argp.parse_args()
+  signal.alarm(args.timeout)
+  def OnResolverResultAvailable(result):
+    answers, authority, additional = result
+    for a in answers:
+      print(a.payload)
```
Comment (dgquintas) [inline test/cpp/naming/test_dns_resolver.py]: this isn't really a test. maybe it should be in some `utils/` directory, or have a different name. Right now it's a bit misleading.

File: test/cpp/naming/test_tcp_connect.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python2.7
```
Comment (dgquintas) [inline test/cpp/naming/test_tcp_connect.py]: ditto wrt to not being really a test, more a util/aux bin.

File: templates/test/cpp/naming/resolver_component_tests_defs.include (hunk 1 above)
Comment (apolcyn) [reply templates/test/cpp/naming/resolver_component_tests_defs.include]: Note the test is already broken under ipv6-only as is (https://github.com/grpc/grpc/blob/master/test/cpp/naming/resolver_component_tests_runner.sh#L168)

File: test/cpp/naming/test_dns_resolver.py (hunk 3 above)
Comment (apolcyn) [reply test/cpp/naming/test_dns_resolver.py]: My personal argument :)  is that the `test_` prefix in the name of this utility is consistent with `test_` prefix being in the name of the already-existing [test_dns_server](https://github.com/grpc/grpc/blob/master/test/cpp/naming/test_dns_server.py) that's in this directory. I could definitely rename both of those to have e.g, `util` in the name, or be under a `util` directory though. WDYT?

File: test/cpp/naming/test_tcp_connect.py (hunk 4 above)
Comment (apolcyn) [reply test/cpp/naming/test_tcp_connect.py]: In the most recent commit I renamed 
`test_dns_server.py` to `dns_server.py`, 
`test_tcp_connect.py` to `tcp_connect.py`,
`test_dns_resolver.py` to `dns_resolver.py`

and moved all three to be under a `utils` subdirectory.

PTAL

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: test/cpp/naming/test_tcp_connect.py (hunk 4 above)
Comment (apolcyn) [reply test/cpp/naming/test_tcp_connect.py]: also added one line forgotten in #14609

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: templates/test/cpp/naming/resolver_component_tests_defs.include (hunk 1 above)
Comment (pmarks-net) [reply templates/test/cpp/naming/resolver_component_tests_defs.include]: Yes, `127.0.0.1` is incompatible with our `ipv6_only` test environment; gRPC is a http://cs/broken_under_ipv6_only bloodbath, but I don't recommend making the problem worse.

Comment (apolcyn) [reply templates/test/cpp/naming/resolver_component_tests_defs.include]: As I understand, the problem is that the DNS server which this script (`$FLAGS_test_dns_resolver_bin_path`) is querying is currently binding to the ipv4 `INADDR_ANY` address ([this python library call to start listening on a port](https://github.com/grpc/grpc/blob/master/test/cpp/naming/test_dns_server.py#L103) uses that by default, [configured around here](https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/tcp.py#L973).

So this `$FLAGS_test_dns_resolver_bin_path` script I believe needs to set it's target address to `127.0.0.1` to be sure to reach it. The same problem exists for the gRPC [test binary](https://github.com/grpc/grpc/blob/master/test/cpp/naming/resolver_component_tests_runner.sh#L168). Neither of these clients can resolve `localhost` too, so I think that fixing this `broken-under ipv6-only` situation for this test would require conditionally checking ipv6 availability for both of the clients and the server.

Would we be ok filing a bug about this test's ipv6-only brokenness and fixing in a separate PR?

**Comment by apolcyn** [comment]
windows test failure: https://github.com/grpc/grpc/issues/14645 (this PR's test is linux/mac only)

File: templates/test/cpp/naming/resolver_component_tests_defs.include (hunk 1 above)
Comment (apolcyn) [reply templates/test/cpp/naming/resolver_component_tests_defs.include]: Filed https://github.com/grpc/grpc/issues/14644 for the ipv6 brokenness

**Pull request merged**